[RFC] Appimage

AdnoC opened the pull request:
Handles https://github.com/neovim/neovim/issues/6083

Command to build the AppImage executable is `make appimage`.

Executable is placed in `build/bin` with naming format ~~`NVim-Nightly-${COMMIT_DATE}-git${GIT_HASH_SUMMARY}.glibc${GLIBC_VERSION}-${ARCHITECTURE}.AppImage`~~.

Name format is now `Neovim-${COMMIT_DATE}-${NEOVIM_VERSION}-glibc${GLIBC_VERSION}-${ARCHITECTURE}.AppImage`.

You can find pre-build executables to test [here](https://github.com/AdnoC/nvim-appimage/releases).

AdnoC committed:
Add AppImage build script

AdnoC committed:
Add "appimage" make target

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
```
It looks like you can simply use `#!/bin/sh`, the only thing which needs to be changed is LOWERAPP. Note that with bash you need `#!/usr/bin/env bash`, other scripts are using that.

ZyX-I replied:
You are using `sh scripts/genappimage.sh` in Makefile. This can’t be `#!/bin/bash` in this case, sh is not bash in Ubuntu, it is dash.

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,143 @@
+  export ARCH="$(arch)"
+fi
+
+APP=NVim
```
Do not remember Neovim being named NVim. Why not Neovim?

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,143 @@
+# get_desktopintegration "$LOWERAPP"
+
+########################################################################
+# Determine the version of the app; also include needed glibc version
```
Where is “needed glibc version”?

AdnoC replied:
Ah, I need to change that comment. It used to add glibc to the version string twice - once there and once in the generate_appimage function in https://github.com/probonopd/AppImages/blob/master/functions.sh

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,143 @@
+fi
+
+APP=NVim
+LOWERAPP="${APP,,}"
```
No need to use weird bash features here, can simply hardcode a string. Though I guess this needs to be nvim and not tolower($APP).

ZyX-I replied:
It also looks like the only place you need bash, except that I do not know what appimage_functions.sh supports.

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,143 @@
+#get_apprun
+
+# get_desktop
+find "${ROOT_DIR}" -name "${LOWERAPP}.desktop" -xdev -exec cp {} "${LOWERAPP}.desktop" \;
```
I can hardly understand what `find` is doing here. Why not simply use a path?

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,143 @@
+
+# Since neovim increments versions slower than vim, using
+# the commit's date makes more sense.
+VIM_VER="$(date -d "@$(git log -1 --format=%ct)" "+%F")"
```
Should be NVIM_VER. And why not *first* include a base version (like 0.2.0) and *then* a date and not only a date? Also `git` knows how to format a date: `git show --no-patch --date='format:%F' --format='%cd'`.

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,143 @@
+
+cd "$APP_BUILD_DIR"
+
+wget -q https://github.com/probonopd/AppImages/raw/master/functions.sh -O ./appimage_functions.sh
```
Our other scripts all use curl. Though most portable would be using `cmake -DURL=https://github.com/probonopd/AppImages/raw/master/functions.sh -DFILE=./appimage_functions.sh cmake/Download.cmake`.

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,143 @@
+#sed -i -e "s|/usr/local/|$APPDIR/usr/|g"     usr/bin/nvim
+#sed -i -e "s|/usr/share/doc/vim/|$APPDIR/usr/share/doc/vim/|g" usr/bin/nvim
+
+patch_strings_in_file "$APP_DIR"/usr/bin/nvim '/usr/local/nvim' '$APPDIR/usr/nvim'
```
Looks hacky, can’t you use some defines for that? CMAKE_INSTALL_PREFIX is defined via `make CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=…"` if you want to use top-level Makefile.

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,143 @@
+# Now packaging it as an AppImage
+########################################################################
+
+find "${ROOT_DIR}" -name "nvim.apprun" -xdev -exec cp {} "$APP_DIR/AppRun" \;
```
Again, why find and not exact path?

ZyX-I commented on appimage/nvim.apprun:
```diff
@@ -0,0 +1,16 @@
+# I don't believe this line is needed since the referenced directories are empty.
+# Uncomment if this changes.
+# Value taken from https://github.com/probonopd/AppImageKit/blob/appimagetool/master/AppRun.c
+# export LD_LIBRARY_PATH="$(printf "LD_LIBRARY_PATH=%s/usr/lib/:%s/usr/lib/i386-linux-gnu/:%s/usr/lib/x86_64-linux-gnu/:%s/usr/lib32/:%s/usr/lib64/:%s/lib/:%s/lib/i386-linux-gnu/:%s/lib/x86_64-linux-gnu/:%s/lib32/:%s/lib64/:%s" "$APPIMAGEDIR" "$APPIMAGEDIR" "$APPIMAGEDIR" "$APPIMAGEDIR" "$APPIMAGEDIR" "$APPIMAGEDIR" "$APPIMAGEDIR" "$APPIMAGEDIR" "$APPIMAGEDIR" "$APPIMAGEDIR" "$LD_LIBRARY_PATH")"
```
This is incorrect, see probonopd/AppImageKit#391.

ZyX-I commented on appimage/nvim.apprun:
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+# This script is a replacement for the standard AppRun binary.
+# The problem is that AppRun normally appends to PYTHONPATH.
```
It does not append, it prepends.

AdnoC committed:
appimage: Clean up build script

Changed generated executable name format. It is now:
NeoVim-Nightly-${COMMIT_DATE}-${NEOVIM_VERSION}-glibc${GLIBC_VERSION}-${ARCHITECTURE}.AppImage

Use the nvim.desktop file already in the runtime directory.
Use symlinks to the nvim executable instead of a wrapper script for AppRun.
Use paths when moving needed files instead of `find`ing them first.
`make install` to the build directory instead of using `DESTDIR` then moving them.
User `curl` instead of `wget`.

Script now shebangs to `/bin/sh` instead of `/bin/bash`.

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,147 @@
+  export ARCH="$(arch)"
+fi
+
+APP=NeoVim
```
Official variant is Neovim. One capital.

AdnoC committed:
appimage: Fix executable name

ZyX-I commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,147 @@
+
+cd "$APP_DIR"
+# No need for a fancy script. AppRun can just be a symlink to nvim.
+ln -s ./usr/bin/nvim AppRun
```
`ln -s` does not really care whether path symlink points to exist. So may just as well be `ln -s usr/bin/nvim "$APP_DIR/AppRun"`, without any `cd`s.

AdnoC replied:
After failing so many symlinks due to using relative paths, I should have realized that by now.

ZyX-I replied:
The cd is necessary in an interactive session: not because ln will start to care, but because tab completion will be enabled and you have less probability of error. This does not apply in scripts.

ZyX-I reviewed (approved):
LGTM, except that I am not familiar with appimage.

AdnoC committed:
appimage: Remove unnecessary cds

AdnoC commented:
I decided to test this on `ubuntu-14.04.5-desktop-amd64.iso` (Trusty). There was one issue due to something in the `functions.sh` script from AppImage using a syntax of function declaration not compatible with `/bin/sh`, but its an easy fix [and I opened a pull request for it](https://github.com/probonopd/AppImages/pull/229).

The generated AppImage ran properly and still worked when ran on `archlinux-2017.04.01-x86_64.iso`. The AppImage built on ArchLinux also ran without errors on Trusty.

ZyX-I commented:
@AdnoC The docker is good for such tests, there are also ubuntu and archlinux images. And travis docker images in https://quay.io/organization/travisci. Unless you are going to mess up with kernel (which is shared with the host) it would normally be faster then whatever you could use iso images for.

ZyX-I commented:
@AdnoC Why merging with master? AFAIK after last policy review you are supposed to

1. Avoid either merging or rebasing if you do not have conflicts or do not need some recent features.
2. For small PRs with easy rebasing (this one is this I think) rebase.
3. Only otherwise merge.

AdnoC commented:
eh?

AdnoC commented:
oh! oops. So thats what "Update branch" does. My bad. I'll force push to remove the merge.

AdnoC commented:
RE: Docker

I hadn't realized that you could run Linux containers on Windows. TIL

justinmk commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,145 @@
+# Determine the version of the app; For use in file filename
+########################################################################
+
+VERSION="Nightly-$COMMIT_DATE-$VIM_VER"
```
remove `Nightly-` please. The version is enough, and for tagged versions we don't want "nightly" in the file name.

justinmk commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,145 @@
+# rm -rf ./usr/share/doc || true
+#rm -rf ./usr/bin/vim || true
+# remove unneded links
+# find ./usr/bin -type l \! -name "gvim" -delete || true
```
is left over from vim's script? Is this a TODO or can we just delete this section?

AdnoC replied:
It was from vim's script and can be deleted

justinmk commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,145 @@
+# #for package in *.deb; do
+# #    dpkg -x $package .
+# #done
+# #rm -f *.deb
```
delete this section starting from line 41

justinmk commented on scripts/genappimage.sh:
```diff
@@ -0,0 +1,145 @@
+#!/bin/sh
+
+########################################################################
+# Package the baries built on Travis-CI as an AppImage
```
typo

AdnoC committed:
appimage: Clean up build script

Removed "Nightly-" from final executable name.
Removed unused and commented out lines from the original vim-appimage script.
Removed unneeded `mv` from `usr/local` to `usr`. We install directly to `usr` now.
Fixed spelling mistake.

justinmk commented:
This is really cool, I tried one of your release images. But when I run `sh scripts/genappimage.sh` or `.  scripts/genappimage.sh` locally, there are errors:

```
ninja: Entering directory `build'
[3/3] Install the project...
FAILED: cd /home/vagrant/neovim/build && /usr/bin/cmake -P cmake_install.cmake
-- Install configuration: "Debug"
-- Installing: /usr/local/share/man/man1/nvim.1
CMake Error at cmake_install.cmake:72 (FILE):
  file INSTALL cannot copy file "/home/vagrant/neovim/man/nvim.1" to
  "/usr/local/share/man/man1/nvim.1".


ninja: build stopped: subcommand failed.
make: *** [install] Error 1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   139  100   139    0     0    261      0 --:--:-- --:--:-- --:--:--   261
100 11364  100 11364    0     0  15789      0 --:--:-- --:--:-- --:--:-- 15789
./scripts/genappimage.sh: 43: cd: can't cd to Neovim.AppDir
./scripts/genappimage.sh: 67: ./scripts/genappimage.sh: [[: not found
./scripts/genappimage.sh: 67: ./scripts/genappimage.sh: [[: not found
./scripts/genappimage.sh: 67: ./scripts/genappimage.sh: [[: not found
./scripts/genappimage.sh: 67: ./scripts/genappimage.sh: [[: not found
./scripts/genappimage.sh: 67: ./scripts/genappimage.sh: [[: not found
./scripts/genappimage.sh: 67: ./scripts/genappimage.sh: [[: not found
ld-linux.so.2 ld-linux-x86-64.so.2 libanl.so.1 libBrokenLocale.so.1 libcidn.so.1 libcrypt.so.1 libc.so.6 libdl.so.2 libm.so.6 libmvec.so.1 libnsl.so.1 libns
s_compat.so.2 libnss_db.so.2 libnss_dns.so.2 libnss_files.so.2 libnss_hesiod.so.2 libnss_nisplus.so.2 libnss_nis.so.2 libpthread.so.0 libresolv.so.2 librt.s
o.1 libthread_db.so.1 libutil.so.1 libstdc++.so.6 libGL.so.1 libdrm.so.2 libxcb.so.1 libX11.so.6 libgio-2.0.so.0 libgdk-x11-2.0.so.0 libgtk-x11-2.0.so.0 lib
asound.so.2 libgdk_pixbuf-2.0.so.0 libfontconfig.so.1 libselinux.so.1 libcom_err.so.2 libcrypt.so.1 libexpat.so.1 libgcc_s.so.1 libglib-2.0.so.0 libgpg-erro
r.so.0 libgssapi_krb5.so.2 # Disputed, seemingly needed by Arch Linux since Kerberos is named differently there libhcrypto.so.4 libhx509.so.5 libICE.so.6 li
bidn.so.11 libk5crypto.so.3 libkeyutils.so.1 libkrb5.so.26 # Disputed, seemingly needed by Arch Linux since Kerberos is named differently there libkrb5.so.3
 # Disputed, seemingly needed by Arch Linux since Kerberos is named differently there libkrb5support.so.0 # Disputed, seemingly needed by Arch Linux since K
erberos is named differently there libp11-kit.so.0 libroken.so.18 libSM.so.6 libusb-1.0.so.0 libuuid.so.1 libwind.so.0 libz.so.1 libgobject-2.0.so.0 libgpg-
error.so.0
ln: failed to access ‘/home/vagrant/neovim/build/Neovim.AppDir/AppRun’: Not a directory
--2017-05-03 21:45:00--  https://github.com/probonopd/AppImageKit/releases/download/6/AppImageAssistant_6-x86_64.AppImage
Resolving github.com (github.com)... 192.30.253.112, 192.30.253.113
Connecting to github.com (github.com)|192.30.253.112|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github-cloud.s3.amazonaws.com/releases/9435153/d4fbde64-7778-11e6-9e10-37eb60fc42de.AppImage?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAISTNZFOVBIJMK3TQ%2F20170504%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170504T095841Z&X-Amz-Expires=300&X-Amz-Signature=ba20a2b09bc1dbc112f9af8fa1b7746664ce8610e974ed86cf0f67906087b138&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3DAppImageAssistant_6-x86_64.AppImage&response-content-type=application%2Foctet-stream [following]
--2017-05-03 21:45:00--  https://github-cloud.s3.amazonaws.com/releases/9435153/d4fbde64-7778-11e6-9e10-37eb60fc42de.AppImage?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAISTNZFOVBIJMK3TQ%2F20170504%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170504T095841Z&X-Amz-Expires=300&X-Amz-Signature=ba20a2b09bc1dbc112f9af8fa1b7746664ce8610e974ed86cf0f67906087b138&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3DAppImageAssistant_6-x86_64.AppImage&response-content-type=application%2Foctet-stream
Resolving github-cloud.s3.amazonaws.com (github-cloud.s3.amazonaws.com)... 52.216.225.160
Connecting to github-cloud.s3.amazonaws.com (github-cloud.s3.amazonaws.com)|52.216.225.160|:443... connected.
HTTP request sent, awaiting response... 416 Requested Range Not Satisfiable

    The file is already fully retrieved; nothing to do.

['/tmp/.mount_YnLLfX/package', './Neovim.AppDir/', '../out/Neovim-v0.2.1-26-g1a09cfe95153.glibc2.17-x86_64.AppImage']
Icon could not be found based on information in desktop file, aborting
mv: cannot stat ‘/home/vagrant/neovim/out/*.AppImage’: No such file or directory
```

Also,  it seems to expect nvim to be built before running. 

I'm trying to gather all the steps needed to automate this on CI.

justinmk commented:
The `]] not found` errors are probably because `sh` on ubuntu is not bash.

justinmk commented:
Looks like AppImageAssistant isn't downloading.

justinmk commented:
> Do not remember Neovim being named NVim. Why not Neovim?

The `APP=Neovim` name is used for the runnable binary name. We teach users to run `nvim`, not `neovim`, so it should be `APP=nvim`. The result will look like `nvim-{version}.AppImage`.

justinmk commented:
Here's where I'm stuck:

```
+ ./AppImageAssistant ./nvim.AppDir/ ../out/nvim-.glibc2.17-x86_64.AppImage
['/tmp/.mount_u65brW/package', './nvim.AppDir/', '../out/nvim-.glibc2.17-x86_64.AppImage']
Icon could not be found based on information in desktop file, aborting
+ mv '/home/vagrant/neovim/out/*.AppImage' /home/vagrant/neovim/build/bin
mv: cannot stat ‘/home/vagrant/neovim/out/*.AppImage’: No such file or directory
```

justinmk commented:
Reordering a `cd` fixed it.

justinmk commented:
@AdnoC Could you try https://github.com/neovim/neovim/pull/6676 on your archlinux machine? On my ubuntu 14.04 machine, 

    ./scripts/genappimage.sh
    ./build/bin/nvim-v0.2.1-26-g1a09cfe95153.glibc2.17-x86_64.AppImage 
    execv error: : No such file or directory

justinmk commented:
Merged. Thank you @AdnoC and @ZyX-I. And thanks to @probonopd -- I'm really impressed by AppImage!

AdnoC commented:
I'm not sure why it would say `]] not found` on Ubuntu 14.04, as I didn't get that when I ran it on a clean install.
